### PR TITLE
[add] Variance Calculation

### PR DESCRIPTION
**Pull request by kingchou007**
variance + efficient nn search

### CODE REVIEW

File: crowndata-evaluation/crowndata_evaluation/services/action_consistency/action_variance_calculator.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+from typing import Dict
+from .clustering import define_clusters
+
+
+class ActionVarianceCalculator:
```
Comment (wxj77) [inline crowndata-evaluation/crowndata_evaluation/services/action_consistency/action_variance_calculator.py:6]: Can you change this to function? easier to maintain and reuse.

File: crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import numpy as np
 from typing import Dict
 import sklearn.cluster
+from scipy.spatial import cKDTree  # or sklearn.neighbors import KDTree
```
Comment (wxj77) [inline crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py]: remove unnecessary comments?

Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py]: - [ ] compare Scipy.spatial and sklearn.neighbors KDTree

File: crowndata-evaluation/crowndata_evaluation/services/action_consistency/action_variance_calculator.py (hunk 1 above)
Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/action_consistency/action_variance_calculator.py:6]: We might do some extensions. I think we can wait until Bobby finishes implementing and then decide whether to delete it or keep it.

File: crowndata-evaluation/crowndata_evaluation/services/utils.py [hunk 3]
```diff
@@ -1,73 +0,0 @@
-import json
-import numpy as np
```
Comment (wxj77) [inline crowndata-evaluation/crowndata_evaluation/services/utils.py]: why remove this file?

File: crowndata-evaluation/crowndata_evaluation/services/__init__.py
Comment (wxj77) [inline crowndata-evaluation/crowndata_evaluation/services/__init__.py:1]: why remove this file?

File: crowndata-evaluation/crowndata_evaluation/services/utils.py (hunk 3 above)
Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/utils.py:2]: rewrite this.

Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/utils.py]: > why remove this file?

I rewrite this func.

File: crowndata-evaluation/crowndata_evaluation/services/__init__.py
Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/__init__.py:1]: new commit will include this file.

Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/__init__.py:1]: > why remove this file?

new commit will include this file.

File: crowndata-evaluation/crowndata_evaluation/services/utils.py (hunk 3 above)
Comment (crowndata) [reply crowndata-evaluation/crowndata_evaluation/services/utils.py]: where is the new function?

File: crowndata-evaluation/crowndata_evaluation/services/__init__.py
Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/__init__.py:1]: https://github.com/crowndata/crowndata-api-python/pull/11/commits/c79e0c5ff8328b9e6fcc7a1ffdff64b36663f2cc

**Comment by kingchou007** [comment]
Screenshot 2024-10-14 at 9 54 57 PM

- [x] Update action variance input to trajectories: Union[Dict[str, np.ndarray], np.ndarray] https://github.com/crowndata/crowndata-api-python/pull/11/commits/96096fe4a392466a817ccaf93d1845fff42d6a1a
- [x] additional variance tests https://github.com/crowndata/crowndata-api-python/pull/11/commits/9f8a8f27f02089584f2bc7dc352da5249f294d78
- [x] Relocate the test to `crowndata-api-python/crowndata-evaluation/tests` https://github.com/crowndata/crowndata-api-python/pull/11/commits/2354278fde14ffc6b70799922cbdee8c3648f60d https://github.com/crowndata/crowndata-api-python/pull/11/commits/31483596e366ef1f39f40c90a7849d5d98c04a39 https://github.com/crowndata/crowndata-api-python/pull/11/commits/11e346d162bbc12d5f0d44c82fc9df07b18b9c3f

File: crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py [hunk 4]
```diff
@@ -47,51 +48,122 @@ def sklearn_cluster_wrapper(
...
+#         The coverage distance for clustering.
+
+#     Returns
+#     -------
+#     Dict[int, np.ndarray]
+#         A dictionary where each key is a cluster index, and the value is the
+#         states or actions in that cluster.
+#     """
+
+#     clusters = {}
```
Comment (crowndata) [inline crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py]: these code are obsolete, can be removed.

**Review by crowndata — CHANGES_REQUESTED** [review]
Can you also change the reference of removed functions in ‘routers’

File: crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py (hunk 4 above)
Comment (kingchou007) [reply crowndata-evaluation/crowndata_evaluation/services/action_consistency/clustering.py]: > these code are obsolete, can be removed.

https://github.com/crowndata/crowndata-api-python/pull/11/commits/e338c025ed72efe72b216120d9de189608f5c1f7

- code clean
- scikit-learn style comment

**Comment by kingchou007** [comment]
> Can you also change the reference of removed functions in ‘routers’

Could you provide a high-level review and Merge this? I will update the 'routers' in the next PR. @wxj77 @crowndata 

https://github.com/crowndata/crowndata-api-python/pull/11#issuecomment-2411359137

**Pull request merged**